README
README for Lingua::EN::Numbers::Ordinate
                                        Time-stamp: "2000-08-24 16:59:23 MDT"

		    Lingua::EN::Numbers::Ordinate

[Partially excerpted from the POD.]


     Lingua::EN::Numbers::Ordinate -- go from cardinal number (3)
     to ordinal ("3rd")

SYNOPSIS
       use Lingua::EN::Numbers::Ordinate;
       print ordinate(4), "\n";
        # prints 4th
       print ordinate(-342), "\n";
        # prints -342nd

       # Example of actual use:
       ...
       for(my $i = 0; $i < @records; $i++) {
         unless(is_valid($record[$i]) {
           warn "The ", ordinate($i), " record is invalid!\n";
           next;
         }
         ...
       }

DESCRIPTION
     There are two kinds of numbers in English -- cardinals (1,
     2, 3...), and ordinals (1st, 2nd, 3rd...).  This library
     provides functions for giving the ordinal form of a number,
     given its cardinal value.

FUNCTIONS
     ordinate(SCALAR)
          Returns a string consisting of that scalar's string
          form, plus the appropriate ordinal suffix.  Example:
          ordinate(23) returns "23rd".

          As a special case, ordinate(undef) and ordinate("")
          return "0th", not "th".

          This function is exported by default.

     th(SCALAR)
          Merely an alias for ordinate, but not exported by
          default.

     ordsuf(SCALAR)
          Returns just the appropriate ordinal suffix for the
          given scalar numeric value.  This is what ordinate uses
          to actually do its work.  For example, ordsuf(3) is
          "rd".

          Not exported by default.

     The above functions are all prototyped to take a scalar
     value, so ordinate(@stuff) is the same as ordinate(scalar
     @stuff).

CAVEATS
     * Note that this library knows only about numbers, not
     number-words.  ordinate('seven') might just as well be
     ordinate('superglue') or ordinate("\x1E\x9A") -- you'll get
     the fallthru case of the input string plus "th".

     * As is unavoidable, ordinate(0256) returns "174th" (because
     ordinate sees the value 174). Similarly, ordinate(1E12)
     returns "1000000000000th".  Returning "trillionth" would be
     nice, but that's an awfully atypical case.

     * Note that this library's algorithm (as well as the basic
     concept and implementation of ordinal numbers) is totally
     language specific.

     To pick a trivial example, consider that in French, 1
     ordinates as "1ier", whereas 41 ordinates as "41ieme".

STILL NOT SATISFIED?
     Bored of this...?

       use Lingua::EN::Numbers::Ordinate qw(ordinate th);
       ...
       print th($n), " entry processed...\n";
       ...

     Try this bit of lunacy:

       {
         my $th_object;
         sub _th () { $th_object }

         package Lingua::EN::Numbers::Ordinate::Overloader;
         my $x; # Gotta have something to bless.
         $th_object = bless \$x; # Define the object now, which _th returns
         use Carp ();
         use Lingua::EN::Numbers::Ordinate ();
         sub overordinate {
           Carp::croak "_th should be used only as postfix!" unless $_[2];
           Lingua::EN::Numbers::Ordinate::ordinate($_[1]);
         }
         use overload '&' => \&overordinate;
       }

     Then you get to do:

       print 3 & _th, "\n";
         # prints "3rd"

       print 1 + 2 & _th, "\n";
         # prints "3rd" too!
         # Because of the precedence of & !

       print _th & 3, "\n";
         # dies with: "th should be used only as postfix!"

     Kooky, isn't it?  For more delightful deleria like this, see
     Damian Conway's Object Oriented Perl from Manning Press.

     Kinda makes you like th(3), doesn't it?

COPYRIGHT
     Copyright (c) 2000 Sean M. Burke.  All rights reserved.

     This library is free software; you can redistribute it
     and/or modify it under the same terms as Perl itself.

AUTHOR
     Sean M. Burke sburke@cpan.org


[End of POD excerpt]

PREREQUISITES

This suite requires Perl 5; I've only used it under Perl 5.004, so for
anything lower, you're on your own.

Lingua::EN::Numbers::Ordinate doesn't use any nonstandard modules.


INSTALLATION

You install Lingua::EN::Numbers::Ordinate, as you would install any
perl module library, by running these commands:

   perl Makefile.PL
   make
   make test
   make install

If you want to install a private copy of Lingua::EN::Numbers::Ordinate
in your home directory, then you should try to produce the initial
Makefile with something like this command:

  perl Makefile.PL LIB=~/perl


DOCUMENTATION

POD-format documentation is included in Ordinate.pm.  POD is readable
with the 'perldoc' utility.  See ChangeLog for recent changes.


MACPERL INSTALLATION NOTES

You needn't bother with the makefiles.  Just make a Lingua directory
in your MacPerl site_lib or lib directory, an EN directory under that,
and a Numbers directory under that, and move Ordinate.pm into there.


SUPPORT

Questions, bug reports, useful code bits, and suggestions for
Lingua::EN::Numbers::Ordinate should just be sent to me at sburke@cpan.org


AVAILABILITY

The latest version of Lingua::EN::Numbers::Ordinate is available from the
Comprehensive Perl Archive Network (CPAN).  Visit
<http://www.perl.com/CPAN/> to find a CPAN site near you.



Changes
Revision history for Perl module Lingua::EN::Numbers::Ordinate

1.05 2021-03-23 NEILB
    - Improved test coverage
    - Tweaked code to not warn on documented behaviour
    - Fixed github repo URL
    - Added mention of Lingua::EN::Numbers's num2en_ordinal
      in SEE ALSO.

1.04_02 2021-03-22 NEILB
    - Sigh. I used // in the changes I made to tests,
      which meant it failed on pre-5.10 perls. Refactored.

1.04_01 2021-03-21 NEILB
    - Extended tests to give full coverage, which prompted:
    - Refactored logic in ordsuf(), which resolves a warning.
      It says it will take cardinals, but then the doc also
      says that if you send nonsense, you'll get back "th"
      as the suffix. So maintained that behaviour.
    - Fixed repo URL to reflect my changed github username

1.04 2015-07-04 NEILB
    - Added [MetaJSON] to dist.ini, so release will include META.json.
      RT#105628 from ETHER++

1.03 2014-05-31 NEILB
    - Switched to Dist::Zilla
    - strict and warnings everywhere, perl min version 5.6.0
    - Added a SEE ALSO section with a couple of modules
    - Added github repo to pod
    - Renamed this file from ChangeLog to Changes,
      reformatted as per CPAN::Changes::Spec

1.02_01 2014-05-17 NEILB
    - Developer release, containing all entries in 1.03, above.

1.02 2004-12-30 SBURKE
    - No code changes, just rebundling
	
0.01 2000-08-25 SBURKE
	- First release to CPAN



lib/Lingua/EN/Numbers/Ordinate.pm
package Lingua::EN::Numbers::Ordinate;

# ABSTRACT: go from cardinal number (3) to ordinal ("3rd")

use 5.006;
use strict;
use warnings;
require Exporter;

our @ISA        = qw/ Exporter  /;
our @EXPORT     = qw/ ordinate  /;
our @EXPORT_OK  = qw/ ordsuf th /;

###########################################################################

=head1 NAME

Lingua::EN::Numbers::Ordinate -- go from cardinal number (3) to ordinal ("3rd")

=head1 SYNOPSIS

  use Lingua::EN::Numbers::Ordinate;
  print ordinate(4), "\n";
   # prints 4th
  print ordinate(-342), "\n";
   # prints -342nd

  # Example of actual use:
  ...
  for(my $i = 0; $i < @records; $i++) {
    unless(is_valid($record[$i]) {
      warn "The ", ordinate($i), " record is invalid!\n"; 
      next;
    }
    ...
  }

=head1 DESCRIPTION

There are two kinds of numbers in English -- cardinals (1, 2, 3...), and
ordinals (1st, 2nd, 3rd...).  This library provides functions for giving
the ordinal form of a number, given its cardinal value.

=head1 FUNCTIONS

=over

=item ordinate(SCALAR)

Returns a string consisting of that scalar's string form, plus the
appropriate ordinal suffix.  Example: C<ordinate(23)> returns "23rd".

As a special case, C<ordinate(undef)> and C<ordinate("")> return "0th",
not "th".

This function is exported by default.

=item th(SCALAR)

Merely an alias for C<ordinate>, but not exported by default.

=item ordsuf(SCALAR)

Returns just the appropriate ordinal suffix for the given scalar
numeric value.  This is what C<ordinate> uses to actually do its
work.  For example, C<ordsuf(3)> is "rd". 

Not exported by default.

=back

The above functions are all prototyped to take a scalar value,
so C<ordinate(@stuff)> is the same as C<ordinate(scalar @stuff)>.


=head1 CAVEATS

* Note that this library knows only about numbers, not number-words.
C<ordinate('seven')> might just as well be C<ordinate('superglue')>
or C<ordinate("\x1E\x9A")> -- you'll get the fallthru case of the input
string plus "th".

* As is unavoidable, C<ordinate(0256)> returns "174th" (because ordinate
sees the value 174). Similarly, C<ordinate(1E12)> returns
"1000000000000th".  Returning "trillionth" would be nice, but that's an
awfully atypical case.

* Note that this library's algorithm (as well as the basic concept
and implementation of ordinal numbers) is totally language specific.

To pick a trivial example, consider that in French, 1 ordinates
as "1ier", whereas 41 ordinates as "41ieme".


=head1 SEE ALSO

L<Lingua::EN::Inflect> provides an C<ORD> function,
which returns the ordinal form of a cardinal number.

L<Lingua::EN::Number::IsOrdinal> provides an C<is_ordinal>
function, which returns true if passed an ordinal number.

L<Lingua::EN::Numbers> provides function C<num2en_ordinal()>
which will take a number and return the ordinal as a word.
So 3 will result in "third".

=head1 REPOSITORY

L<https://github.com/neilb/Lingua-EN-Numbers-Ordinate>

=head1 COPYRIGHT

Copyright (c) 2000 Sean M. Burke.  All rights reserved.

This library is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 AUTHOR

Sean M. Burke C<sburke@cpan.org>

This has been maintained by Neil Bowers (NEILB)
since 2014.

=cut

###########################################################################

sub ordsuf ($) {
  return 'th' if not(defined($_[0])) or $_[0] !~ /^-?[0-9]+$/;
   # 'th' for undef, 0, or anything non-number.
  my $n = abs($_[0]);  # Throw away the sign.

  $n %= 100;
  return 'th' if $n == 11 or $n == 12 or $n == 13;
  $n %= 10;
  return 'st' if $n == 1; 
  return 'nd' if $n == 2;
  return 'rd' if $n == 3;
  return 'th';
}

sub ordinate ($) {
  my $i = $_[0] || 0;
  return $i . ordsuf($i);
}

no warnings 'all';
*th = \&ordinate; # correctly copies the prototype, too.

###########################################################################
1;

__END__
